src/libs/antares/study/parallel-years.cpp

```diff
@@ -0,0 +1,431 @@
+        numberOfMCYearThreads[ncMax] = nbLogicalCores - 1;
+        break;
+    }
+
+    /*
+            Getting the number of parallel years based on the number
+            of cores level.
+            This number is limited by the smallest refresh span (if at least
+            one type of time series is generated)
+    */
+
+    try
+    {
+        forcedNbOfParallelYears_ = numberOfMCYearThreads.at(p.nbCores.ncMode);
+    }
+    catch(const std::out_of_range& e)
+    {
+        logs.fatal() << "Simulation cores level not correct : " << (int)p.nbCores.ncMode;
+    }
+    
+}
+
+
+void SetsOfParallelYearCalculator::limitNbOfParallelYearsbyMinRefreshSpan()
+{
+    uint TSlimit = UINT_MAX;
+    if ((p.timeSeriesToGenerate & timeSeriesLoad) && (p.timeSeriesToRefresh & timeSeriesLoad))
+        TSlimit = p.refreshIntervalLoad;
+    if ((p.timeSeriesToGenerate & timeSeriesSolar) && (p.timeSeriesToRefresh & timeSeriesSolar))
+        TSlimit = std::min(p.refreshIntervalSolar,  TSlimit);
+    if ((p.timeSeriesToGenerate & timeSeriesHydro) && (p.timeSeriesToRefresh & timeSeriesHydro))
+        TSlimit = std::min(p.refreshIntervalHydro, TSlimit);
+    if ((p.timeSeriesToGenerate & timeSeriesWind) && (p.timeSeriesToRefresh & timeSeriesWind))
+        TSlimit = std::min(p.refreshIntervalWind, TSlimit);
+    if ((p.timeSeriesToGenerate & timeSeriesThermal) && (p.timeSeriesToRefresh & timeSeriesThermal))
+        TSlimit = std::min(p.refreshIntervalThermal, TSlimit);
+
+    forcedNbOfParallelYears_ = std::min({p.nbYears, TSlimit, forcedNbOfParallelYears_});
+}
+
+bool SetsOfParallelYearCalculator::isRefreshNeededForCurrentYear(uint y)
+{
+    bool refreshing = false;
+    refreshing = (p.timeSeriesToGenerate & timeSeriesLoad)
+                    && (p.timeSeriesToRefresh & timeSeriesLoad)
+                    && ((y % p.refreshIntervalLoad) == 0);
+    refreshing = refreshing
+                    || ((p.timeSeriesToGenerate & timeSeriesSolar)
+                        && (p.timeSeriesToRefresh & timeSeriesSolar)
+                        && (y % p.refreshIntervalSolar) == 0);
+    refreshing = refreshing
+                    || ((p.timeSeriesToGenerate & timeSeriesWind)
+                        && (p.timeSeriesToRefresh & timeSeriesWind)
+                        && (y % p.refreshIntervalWind) == 0);
+    refreshing = refreshing
+                    || ((p.timeSeriesToGenerate & timeSeriesHydro)
+                        && (p.timeSeriesToRefresh & timeSeriesHydro)
+                        && (y % p.refreshIntervalHydro) == 0);
+
+    bool haveToRefreshTSThermal 
+        = ((p.timeSeriesToGenerate & timeSeriesThermal)
+                        && (p.timeSeriesToRefresh & timeSeriesThermal)) || thermalTSRefresh_;
+    refreshing
+        = refreshing || (haveToRefreshTSThermal && (y % p.refreshIntervalThermal == 0));
+
+    return refreshing;
+
+}
+
+void SetsOfParallelYearCalculator::buildSetsOfParallelYears()
```

Comment:
I think we could make this algorithm drier, and reject some data member assignement in a further method.
For instance, `set->nbYears` can be computed later, as well as `set->yearFailed` or 
`set->isFirstPerformedYearOfASet` (not exhaustive list).
It would clarify this algorithm.